dimod/higherorder/utils.py

```diff
@@ -269,7 +269,7 @@ def var(x):
     return cqm
 
 
-def make_quadratic(poly: Mapping[Tuple[Variable, Variable], Bias], strength: float,
+def make_quadratic(poly: Polynomial, strength: float,
```

Comment:
Isn't this also `Union[Polynomial, BinaryPolynomial]`? Applies other places as well

Reply:
Is not currently stated as such: https://docs.ocean.dwavesys.com/en/stable/docs_dimod/reference/generated/dimod.higherorder.utils.make_quadratic.html#dimod.higherorder.utils.make_quadratic 
Might just have been missed. I can check that it works and add, assuming we don't want to only allow the one in some cases

Reply:
Anywhere we do `poly = _init_binary_polynomial(poly, vartype)` it seems to work

Reply:
Is this not a worry though?

```
>>> poly = dimod.BinaryPolynomial({"a": -2, ("a", "b"): -0.5, ("b", "c"): 1.5, ("a", "b", "c"): 3}, dimod.SPIN)
>>> dimod.poly_energy({"a": 1, "b": 1, "c": -1}, poly)
-7.0
>>> dimod.poly_energy({"a": 1, "b": 0, "c": 0}, poly)
-2.0
```

For just `poly: Polinomial` it should be the user's responsibility to ensure that sample vartype is correct for the poly but if we allow `BinaryPolynomial` shouldn't the energy calculation balk at a mismatched vartype?

Reply:
Ah, interesting. Probably `_init_binary_polynomial` should handle that case.

Reply:
Changing `return BinaryPolynomial(poly, 'SPIN').energies(samples_like)` to use `_init_binary_polynomial` won't help with the static `poly_energy` & `poly_energies` methods. And I think maybe this should be checked in `dimod.sampleset.as_samples`: 
```
>>> poly_energy({"a": 1, "b": -1, "c": 0}, poly)
-1.5
```
I think `as_samples` should barf at a "sample" that has both 0 and -1 in it.

Reply:
We are actually very permissive for samples, deliberately. We've considered adding a `strict=True` (https://github.com/dwavesystems/dimod/issues/954) keyword on occasion, but IMO it's better to assume the user has a reason.

Reply:
Okay, I think I added `BinaryPolynomial` to all relevant functions